### DISCUSSION TITLE
Field interpolation

### DISCUSSION THREAD

**Discussion by GenB31415** [zwicker-group/py-pde#104] [Q&A]
Hello everyone!
In my pde-equation problem, I calculate some localized state, which I then want to distribute (copy) to different point(s) of the original space (grid). Then I will use this multi-state as a new initial state. In Matlab's case, I can capture such a localized state and then copy it to different places as some kind of interpolation task using the "pdeInterpolant" function. Can I use the "Field Interpolation" function from very promising py-pde for this? Thank you.

**Comment by david-zwicker** [comment]
Would the method [`interpolate_to_grid`](https://py-pde.readthedocs.io/en/latest/packages/pde.fields.html?highlight=interpolate#pde.fields.base.DataFieldBase.interpolate_to_grid) be helpful?

**Reply by GenB31415** [reply]
I tried to use the method "interpolate_to_grid", but something goes wrong and the exception occurs "AttributeError: 'CartesianGrid' object has no attribute 'interpolate to grid'". I appreciate for a help to indicate how to correct this. Thank you in advance.

My code
import numpy as np
from pde import CartesianGrid, ScalarField
L=10; W=2
gridOld = CartesianGrid([[0, 2 * L]] * 2, 64) #initial 2D grid
myFun = "exp(-((x-" + str(L) + ")**2+(y-" + str(L)+")**2)/(2*" + str(W) + "**2))" 
print("\nmyFun = "+myFun+"\n")
stateOld = ScalarField.from_expression(gridOld, myFun) # <--- old state
stateOld.plot()
# Now try to create the interpolator to new grid
gridNew = CartesianGrid([[0, 2 * L]] * 2, 2*64) # <--- new grid - - -
# and interpolate myFun to the new grid:
fieldCollNew = gridNew.interpolate_to_grid(gridNew, "scipy", "linear", "New state") 
# But Error: ---> AttributeError: 'CartesianGrid' object has no attribute 'interpolate_to_grid'
fieldCollNew.data.plot()

**Reply by david-zwicker** [reply]
Note that the method is defined for fields, not grids.

**Comment by GenB31415** [comment]
Thank you.
    On Tuesday, May 11, 2021, 10:30:24 AM CDT, David Zwicker ***@***.***> wrote:  
 
 


Note that the method is defined for fields, not grids.

**Reply by GenB31415** [reply]
Thanks for the advice. Now the code runs correctly.  So the question is ### Could I indicate the insertion point(s) [x,y] of the interpolated data to new grid? ### Thank you,

My code:

from matplotlib.pyplot import fill
import numpy as np
from pde import CartesianGrid, ScalarField, FieldCollection

L, W = 10, 2
gridOld = CartesianGrid([[0, 2 * L]] * 2, 64) #initial 2D grid
myFun = "exp(-((x-" + str(L) + ")**2+(y-" + str(L)+")**2)/(2*" + str(W) + "**2))" 
print("\nmyFun = "+myFun+"\n")
stateOld = ScalarField.from_expression(gridOld, myFun) # old state
stateOld.plot()
fieldCollectionOld = FieldCollection(fields = [stateOld])
gridNew = CartesianGrid([[0, 1.5 * L],[0, 1.5 * L]] , 2*64) # create new 2D grid - - -
print("Create the interpolator (please wait)...")
fieldCollNew = fieldCollectionOld.interpolate_to_grid(gridNew)
print("Done\n")
print("Plotting new grid:")
fieldCollNew.plot()
print("\nFinish\n")

**Comment by GenB31415** [comment]
I found simple solution that is good for my problem

My code:
from pde import CartesianGrid, ScalarField
L, W = 10, 3 #parameters of Gaussian
myFun = "exp(-((x-" + str(L) + ")**2+(y-" + str(L)+")**2)/(2*" + str(W) + "**2))" 
print("\n"+"myFun = " + myFun )
gridOld = CartesianGrid([[0, 2 * L]] * 2, 64) #initial 2D grid
stateOld = ScalarField.from_expression(gridOld, myFun, label="Init state") # old state
stateOld.plot()
dataOld = stateOld.data
gridNew = CartesianGrid([[0, 1.8 * L],[0, 1.8 * L]] , 2*64) # create new 2D grid 
print("Create new 2x2 state:")
fieldNew = ScalarField(gridNew, 1, label="Final 2x2 state")
for r in range(0,64+1,64):
    for c in range(0,64+1,64):
        fieldNew.data[r:r+dataOld.shape[0], c:c+dataOld.shape[1]] = dataOld
fieldNew.plot()
print("Finish\n")

**Reply by david-zwicker** [reply]
That's a good solution. I guess we cannot supply a method for each special requirement, but since you can simply access the field data as a numpy array, you can certainly create your own solutions. Good job!

**Comment by GenB31415** [comment]
Thank you!
    On Friday, May 14, 2021, 12:30:56 PM CDT, David Zwicker ***@***.***> wrote:  
 
 


That's a good solution. I guess we cannot supply a method for each special requirement, but since you can simply access the field data as a numpy array, you can certainly create your own solutions. Good job!
